File: 백준/Silver/2512. 예산/예산.c
#include <stdio.h>
#include <stdlib.h>

int cmp(const void* a, const void* b) {
	if (*(int*)a > *(int*)b) {
		return 1;
	}
	else if (*(int*)a < *(int*)b) {
		return -1;
	}
	else {
		return 0;
	}
}

int main(void) {
	int n, m, i, sum = 0, right, left, mid;
	int* an;
	scanf("%d", &n);
	an = (int*)malloc(n * sizeof(int));

	for (i = 0; i < n; i++) {
		scanf("%d", &an[i]);
		sum += an[i];
	}

	scanf("%d", &m);

	qsort(an, n, sizeof(int), cmp);

	if (sum <= m) {
		printf("%d", an[n - 1]);
	}
	else {
		left = 0;
		right = an[n - 1];
		while (1) {
			mid = (left + right) / 2;
			sum = 0;
			for (i = 0; i < n; i++) {
				if (an[i] <= mid) {
					sum += an[i];
				}
				else {
					sum += mid;
				}
			}
			if (sum == m) {
				printf("%d", mid);
				break;
			}
			else if (left >= right) {
				if (sum > m) {
					printf("%d", mid - 1);
					break;
				}
				else {
					printf("%d", mid);
					break;
				}
			}
			else if (sum < m) {
				left = mid + 1;
			}
			else {
				right = mid - 1;
			}
		}
	}

	free(an);
	return 0;
}

File: 백준/Silver/2512. 예산/예산.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
	public static void main(String[] args) throws IOException {
		int n;
		int i;
		int max;
		int thr;
		int sum;
		int mid;
		int left;
		int right;
		int[] arr;
		BufferedReader br;
		StringTokenizer st;

		br = new BufferedReader(new InputStreamReader(System.in));
		n = Integer.parseInt(br.readLine());
		arr = new int[n];
		max = 0;
		sum = 0;
		st = new StringTokenizer(br.readLine(), " ", false);
		for (i = 0; i < n; i++) {
			arr[i] = Integer.parseInt(st.nextToken());
			max = Math.max(max, arr[i]);
			sum += arr[i];
		}
		thr = Integer.parseInt(br.readLine());
		if (sum <= thr) {
			System.out.print(max);
			return;
		}
		left = 1;
		right = thr + 1;
		loop:
		while (left < right) {
			mid = left + right >> 1;
			sum = 0;
			for (i = 0; i < n; i++) {
				if ((sum += Math.min(arr[i], mid)) > thr) {
					right = mid;
					continue loop;
				}
			}
			left = mid + 1;
		}
		System.out.print(right - 1);
	}
}
